File: pointers-references/Cat.cpp
#include <iostream>
using namespace std;

struct Cat
{
    string name;
};

void pass_by_reference(Cat &c_ref)
{
    cout << "c_ref name (before): " << c_ref.name << "\n";
    c_ref.name = "Nermal";
    cout << "c_ref name (after): " << c_ref.name << "\n";
}

void pass_by_value(Cat c_val)
{
    cout << "c_val name (before): " << c_val.name << "\n";
    c_val.name = "Socks";
    cout << "c_val name (after): " << c_val.name << "\n";
}

int main()
{
    Cat cat1;
    cat1.name = "Garfield";

    cout << "cat1 name (before): " << cat1.name << "\n";
    pass_by_reference(cat1);
    cout << "cat1 name (after): " << cat1.name << "\n";

    Cat cat2;
    cat2.name = "Tom";

    cout << "cat2 name (before): " << cat2.name << "\n";
    pass_by_value(cat2);
    cout << "cat2 name (after): " << cat2.name << "\n";

    return 0;
}

File: pointers-references/Cat.java

public class Cat {

    public Cat(String name) {
        super();
        this.name = name;
    }
    String name;

    public static void passByValue(Cat c) {
        System.out.println("c name (before): " + c.name);
        c = new Cat("Nermal");
        System.out.println("c name (after): " + c.name);
    }

    public static void actOnValue(Cat c2) {
        System.out.println("c2 name (before): " + c2.name);
        c2.name = "Tom";
        System.out.println("c2 name (after): " + c2.name);
    }

    public static void main(String[] args) {
        Cat cat1 = new Cat("Garfield");

        System.out.println("Cat 1 name (before): " + cat1.name);
        passByValue(cat1);
        System.out.println("Cat 1 name (after): " + cat1.name);

        Cat cat2 = new Cat("Socks");

        System.out.println("Cat 2 name (before): " + cat2.name);
        actOnValue(cat2);
        System.out.println("Cat 2 name (after): " + cat2.name);
    }

}
